File: docs/JAMIA/comparisons/RPython/TM/48062499_string_as_numeric.py
# ID: SO#48062499
# Language: Python
# Bug class: Value type
# Runtime consequence: IF
# In study: Yes
# Inclusion rationale: Y-axis data plotted as strings, not sorted numerically. String where number expected. Data processing error visible in output.
# Reproduction status: Live
# Type system catch: `Record<string, number>` rejects string column

import numpy as np

# Simulates reading a CSV column without type parsing — values are strings
# (exact pattern from the SO question: line[1] from split CSV)
Solar = ["50.35", "41.01", "69.16", "94.5", "111.9",
         "103", "98.6", "36.45", "34.74", "34.17", "34.6"]

# argsort on string array — sorts lexicographically, not numerically
order = np.argsort(Solar)
sorted_solar = np.array(Solar)[order]
print("Sorted as strings (WRONG — silent):")
print(sorted_solar)
# ['103' '111.9' '34.17' '34.6' '34.74' '36.45' '41.01' '50.35' '69.16' '94.5' '98.6']
# "103" comes before "34" — this is lexicographic, not numeric.
# No error, no warning. The plot just looks wrong.
print()

# The fix: explicit type conversion
Solar_f = [float(x) for x in Solar]
order_f = np.argsort(Solar_f)
sorted_f = np.array(Solar_f)[order_f]
print("Sorted as floats (correct):")
print(sorted_f)

File: docs/JAMIA/comparisons/RPython/TM/48062499_string_as_numeric.ts
/**
 * RPython SO#48062499 — numeric data arrives as strings, sorts lexicographically
 *
 * In pandas/numpy, a string array sorts as "103" < "34.17" (lexicographic)
 * with no error. Plots render with wrong y-axis order silently.
 *
 * In tidy-ts, if a column is string, passing it to a numeric operation
 * (like s.mean or s.test) is a compile-time error. The type system forces
 * explicit conversion before numeric use.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

// Data arrives as strings (simulating CSV read without type parsing)
const df = createDataFrame([
  { time: "07:00", solar: "50.35" },
  { time: "08:00", solar: "41.01" },
  { time: "09:00", solar: "69.16" },
  { time: "10:00", solar: "94.5" },
  { time: "11:00", solar: "111.9" },
  { time: "12:00", solar: "103" },
]);

// Schema is { time: string, solar: string }
// Attempting numeric operations on string column fails at compile time:
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.mean(df.extract("solar"));

// The fix: explicit conversion — no silent lexicographic sorting possible
const numeric = df.mutate({ solar: (r) => parseFloat(r.solar) });
// Now solar is number — arrange and stats work correctly
const sorted = numeric.arrange("solar", "asc");
const avg = s.mean(numeric.extract("solar"));
console.log(avg);
